# PRAKTIKUM

In [1] Import library dan memuat Iris. Library dasar di-import dan dataset Iris dimuat dari Scikit-Learn.

In [1]:
import numpy as np
import pandas as pd
from sklearn.datasets import load_iris

# Load dataset Iris dari scikit-learn
iris = load_iris()

In [2] Konversi ke DataFrame. Data diubah menjadi DataFrame. X berisi 4 fitur numerik (sepal length, sepal width, petal length, petal width), y berisi nama spesies.

In [2]:
# Konversi data menjadi DataFrame Pandas
df = pd.DataFrame(iris.data, columns=iris.feature_names)
df["species"] = iris.target_names[iris.target]

# Definisikan X (fitur) dan y (target/label)
X = df.drop(columns=["species"])
y = df["species"]

print(X)

     sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                  5.1               3.5                1.4               0.2
1                  4.9               3.0                1.4               0.2
2                  4.7               3.2                1.3               0.2
3                  4.6               3.1                1.5               0.2
4                  5.0               3.6                1.4               0.2
..                 ...               ...                ...               ...
145                6.7               3.0                5.2               2.3
146                6.3               2.5                5.0               1.9
147                6.5               3.0                5.2               2.0
148                6.2               3.4                5.4               2.3
149                5.9               3.0                5.1               1.8

[150 rows x 4 columns]


In [3] Melihat target (y). Kolom species ditampilkan, berisi 150 data dengan tiga kelas: setosa, versicolor, virginica.

In [3]:
y

0         setosa
1         setosa
2         setosa
3         setosa
4         setosa
         ...    
145    virginica
146    virginica
147    virginica
148    virginica
149    virginica
Name: species, Length: 150, dtype: str

In [4] Encode label. LabelEncoder mengubah nama spesies menjadi angka: setosa = 0, versicolor = 1, virginica = 2, agar bisa diproses model.

In [4]:
from sklearn.preprocessing import LabelEncoder
lab = LabelEncoder()
y = lab.fit_transform(y)
y

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

In [5] Pembagian data. Data dibagi 80% latih (120 data) dan 20% uji (30 data) dengan train_test_split.

In [5]:
from sklearn.model_selection import train_test_split
X_train, X_test = train_test_split(X, test_size=0.2, random_state=0)
y_train, y_test = train_test_split(y, test_size=0.2, random_state=0)

In [6] Melatih model Gaussian Naïve Bayes. Karena semua fitur Iris bertipe numerik kontinu, digunakan Gaussian Naïve Bayes, yang mengasumsikan tiap fitur mengikuti distribusi normal pada setiap kelas. Model dilatih dengan fit(X_train, y_train).

In [6]:
from sklearn.naive_bayes import GaussianNB
model = GaussianNB()
NB_model = model.fit(X_train, y_train)

In [7] Prediksi data uji. Model memprediksi kelas untuk 30 data uji.

In [7]:
result = NB_model.predict(X_test)
result

array([2, 1, 0, 2, 0, 2, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 0, 2, 1,
       0, 0, 2, 0, 0, 1, 1, 0])

In [8] Evaluasi akurasi. Akurasi = 96,67% (29 dari 30 data uji benar), menunjukkan Gaussian Naïve Bayes bekerja sangat baik pada dataset Iris.

In [8]:
from sklearn.metrics import accuracy_score
print('Accuracy of data testing: ', accuracy_score(y_test, result)*100)

Accuracy of data testing:  96.66666666666667


In [9] Prediksi input baru (versi awal). Input [[5.1, 3.5, 1.4, 0.2]] diprediksi sebagai setosa. Namun muncul UserWarning karena input berupa list/array biasa tanpa nama kolom, sedangkan model dilatih dengan DataFrame yang punya nama kolom (iris.feature_names).

In [9]:
#prediksi inputan data baru
#input_data = pd.DataFrame(np.array([[5.1, 3.5, 1.4, 0.2]]), columns=iris.feature_names)
input_data = [[5.1, 3.5, 1.4, 0.2]]
result = NB_model.predict(input_data)
prediction_label = lab.inverse_transform(result)
print('Hasil prediksi untuk input baru adalah:', prediction_label)

Hasil prediksi untuk input baru adalah: ['setosa']


c:\Users\personal\anaconda3\Lib\site-packages\sklearn\utils\validation.py:2827: UserWarning: X does not have valid feature names, but GaussianNB was fitted with feature names
  warnings.warn(


In [10] Prediksi input baru (versi diperbaiki). Input diubah menjadi DataFrame dengan nama kolom yang sama seperti data latih, sehingga peringatan hilang. Hasilnya tetap setosa, karena hanya struktur datanya yang diperbaiki, bukan nilainya.

In [10]:
input_data = pd.DataFrame(np.array([[5.1, 3.5, 1.4, 0.2]]), columns=iris.feature_names)
result = NB_model.predict(input_data)
prediction_label = lab.inverse_transform(result)
print('Hasil prediksi untuk input baru adalah:', prediction_label)


Hasil prediksi untuk input baru adalah: ['setosa']


# TUGAS

Mengikuti cara pada In [10] (memakai DataFrame dengan nama kolom, agar tidak muncul peringatan):

In [11]:
input_data = pd.DataFrame(np.array([ [7.3, 3.8, 4.6, 2.3]]), columns=iris.feature_names)
result = NB_model.predict(input_data)
prediction_label = lab.inverse_transform(result)
print('Hasil prediksi untuk input baru adalah:', prediction_label)

Hasil prediksi untuk input baru adalah: ['virginica']


Hasil: ['virginica']

Model sangat yakin dengan prediksi ini — probabilitas untuk kelas virginica sebesar 1,0 (100%), sedangkan setosa dan versicolor mendekati 0. Hal ini wajar karena nilai petal length (4,6 cm) dan petal width (2,3 cm) pada input ini berada pada rentang nilai yang lebih dekat dengan data virginica dibanding dua kelas lainnya.



Kesimpulan

Model Gaussian Naïve Bayes yang dilatih pada dataset Iris mencapai akurasi 96,67% pada data uji, menunjukkan performa yang sangat baik untuk klasifikasi tiga spesies bunga. Prediksi terhadap input baru [7.3, 3.8, 4.6, 2.3] menghasilkan kelas virginica dengan tingkat keyakinan yang sangat tinggi, karena ciri-ciri fisik pada input tersebut (terutama ukuran petal) paling menyerupai karakteristik bunga virginica dibanding setosa atau versicolor.